### PR TITLE
Tenure extend when the previous tenure is bad

### LINKED ISSUE #5361: Signer should allow a tenure-extend to let the previous miner continue mining if the next miner did not start mining before the timeout
Signers should accept tenure extends from a previous miner if the current miner did not start mining before the required timeout.

### PR DESCRIPTION
**Pull request by obycode**
### Description
Updates the miner to attempt to continue its tenure if the winner of the next tenure commits to the wrong block or if it mines no blocks within some grace period.

### Applicable issues

- fixes #5361

### Checklist

- [ ] Test coverage for new or modified code paths
- [ ] Changelog is updated
- [ ] Required documentation changes (e.g., `docs/rpc/openapi.yaml` and `rpc-endpoints.md` for v2 endpoints, `event-dispatcher.md` for new events)
- [ ] New clarity functions have corresponding PR in `clarity-benchmarking` repo
- [ ] New integration test(s) added to `bitcoin-tests.yml`

### CODE REVIEW

**Comment by obycode** [comment]
Ok, the case where the winning miner has committed to the wrong parent tenure is complete. Now I just need to figure out a clean way to attempt a tenure extend after some time has passed and the new miner has not mined a block.

**Comment by obycode** [comment]
There is a problem that I discovered which is unrelated to these changes, but becomes a challenge due to these changes.
https://github.com/stacks-network/stacks-core/blob/6e0bd5ae66b65797e76b3a48fdf99be22f73459f/testnet/stacks-node/src/event_dispatcher.rs#L110-L123

This global becomes a problem when we have tests that run multiple miners. If multiple miners attempt to propose a block at the same time, then we end up having problems.

**Comment by kantai** [comment]
> This global becomes a problem when we have tests that run multiple miners. If multiple miners attempt to propose a block at the same time, then we end up having problems.

The only clear option to me is to make this a Arc+Mutex an instance variable in the `EventObserver` struct. This requires that the `EventObserver` only be _constructed_ once, and then cloned whenever copies are needed. I think that wouldn't be too hard (there's already a lot of passing around of the object I think, rather than it being continuously reconstructed from the config).

**Comment by obycode** [comment]
This PR handles the tenure extend for the case where the next miner has committed to the wrong parent tenure. The other case mentioned in #5361, where a tenure extend is allowed after some time limit has passed with no blocks from the new miner, will be completed in a separate PR.

**Comment by hstove** [comment]
@obycode can we create a new issue for the timeout case, and then rename #5361 to be more about "if the new miner commits to the wrong block"? Or some derivation of that?

File: stacks-signer/src/signerdb.rs [hunk 1]
```diff
@@ -600,6 +640,15 @@ impl SignerDb {
         try_deserialize(result)
     }
 
+    /// Return the canonical tip -- the last globally accepted block.
```
Comment (jferrant) [inline stacks-signer/src/signerdb.rs:643]: Mind adding a unit test for this?

File: testnet/stacks-node/src/nakamoto_node/relayer.rs [hunk 2]
```diff
@@ -880,29 +905,79 @@ impl RelayerThread {
...
+            NakamotoNodeError::ParentNotFound
+        })?;
+        let highest_tenure_bhh =
+            BlockHeaderHash(highest_tenure_start_block_header.index_block_hash().0);
+
+        // If we won the last good sortition, then we should extend off of it.
+        let last_good_block_election_snapshot = {
             let ih = self.sortdb.index_handle(&burn_tip.sortition_id);
-            ih.get_last_snapshot_with_sortition(burn_tip.block_height)
+            info!(
```
Comment (jferrant) [inline testnet/stacks-node/src/nakamoto_node/relayer.rs:931]: Should these be debug logs?

File: stacks-signer/src/signerdb.rs [hunk 3]
```diff
@@ -405,9 +415,24 @@ static SCHEMA_3: &[&str] = &[
     "INSERT INTO db_config (version) VALUES (3);",
 ];
 
+static SCHEMA_4: &[&str] = &[
```
Comment (hstove) [inline stacks-signer/src/signerdb.rs:418]: Check out the way I did schema changed in my PR around retrying 429'd verifications: https://github.com/stacks-network/stacks-core/pull/5453/files#diff-2142a830df37bd7d21338d998aa3c4486f0ccbab7476254db18997ef49a463c3

I don't think we should be dropping tables unless necessary, and since we're just adding an index and changing db_version, this PR can do the same approach

File: testnet/stacks-node/src/tests/signer/v0.rs [hunk 4]
```diff
@@ -7037,3 +7037,832 @@ fn block_validation_response_timeout() {
...
+
+#[test]
+#[ignore]
+/// Test that a miner will extend its tenure after the succeding miner commits to the wrong block.
+/// - Miner 1 wins a tenure and mines normally
+/// - Miner 1 wins another tenure and mines normally, but miner 2 does not see any blocks from this tenure
+/// - Miner 2 wins a tenure and is unable to mine a block
+/// - Miner 1 extends its tenure and mines an additional block
+/// - Miner 2 wins the next tenure and mines normally
+fn tenure_extend_after_bad_commit() {
```
Comment (jferrant) [inline testnet/stacks-node/src/tests/signer/v0.rs:7440]: Nice test! Was easy to follow.

File: testnet/stacks-node/src/nakamoto_node/relayer.rs [hunk 5]
```diff
@@ -880,29 +905,79 @@ impl RelayerThread {
...
+            }
         };
 
-        let won_last_sortition = last_block_election_snapshot.miner_pk_hash == Some(mining_pkh);
-        debug!(
-            "Relayer: Current burn block had no sortition. Checking for tenure continuation.";
+        let won_last_sortition =
+            last_good_block_election_snapshot.miner_pk_hash == Some(mining_pkh);
+        info!(
+            "Relayer: Current burn block had no sortition or a bad sortition. Checking for tenure continuation.";
```
Comment (hstove) [inline testnet/stacks-node/src/nakamoto_node/relayer.rs:969]: Is this true? Looking at line 962, the case where `sn.winning_stacks_block_hash == highest_tenure_bhh` - wouldn't that mean the current burn block has a good sortition? Or is the fact that we're in `continue_tenure` mean that my assertion is wrong? Sorry, it's tricky to follow all the code paths here.

File: testnet/stacks-node/src/nakamoto_node/relayer.rs [hunk 6]
```diff
@@ -880,29 +905,79 @@ impl RelayerThread {
...
                     NakamotoNodeError::SnapshotNotFoundForChainTip
+                })?;
+            if sn.winning_stacks_block_hash != highest_tenure_bhh {
+                info!(
+                    "Relayer: Sortition winner is not committed to the canonical tip; allowing last miner to extend";
+                    "burn_block_height" => burn_tip.block_height,
+                    "consensus_hash" => %burn_tip.consensus_hash,
+                );
+
+                SortitionDB::get_block_snapshot_consensus(
```
Comment (hstove) [inline testnet/stacks-node/src/nakamoto_node/relayer.rs:949]: At line 984 we run this exact query. With your change, is that always necessary? If it's maybe necessary, can we change 984 to something like:

```rust
let canonical_block_snapshot = if last_good_block_election_snapshot
    .canonical_stacks_tip_consensus_hash
    == canonical_stacks_tip_ch
{
    last_good_block_election_snapshot
} else {
    SortitionDB::get_block_snapshot_consensus(self.sortdb.conn(), &canonical_stacks_tip_ch)
        .map_err(|e| {
            error!("Relayer: failed to get block snapshot for canonical tip: {e:?}");
            NakamotoNodeError::SnapshotNotFoundForChainTip
        })?
        .ok_or_else(|| {
            error!("Relayer: failed to get block snapshot for canonical tip");
            NakamotoNodeError::SnapshotNotFoundForChainTip
        })?
};
```

File: testnet/stacks-node/src/tests/signer/v0.rs [hunk 7]
```diff
@@ -420,6 +420,22 @@ impl SignerTest<SpawnedSigner> {
     }
 }
 
+fn verify_last_block_contains_tenure_change_tx(cause: TenureChangeCause) {
```
Comment (hstove) [inline testnet/stacks-node/src/tests/signer/v0.rs:423]: Nit: can you change line 2806 to use this function? I guess you might need to refactor the function to not use `panic`/`assert_eq`. Perhaps the best change would be to have `last_block_contains_tenure_change -> bool`, and then also `verify_last_block_contains_tenure_change`.

File: testnet/stacks-node/src/tests/signer/v0.rs [hunk 8]
```diff
@@ -7037,3 +7037,832 @@ fn block_validation_response_timeout() {
...
+    let burn_height_before = get_burn_height();
+
+    next_block_and(
+        &mut signer_test.running_nodes.btc_regtest_controller,
+        60,
+        || Ok(get_burn_height() > burn_height_before),
+    )
+    .expect("Timed out waiting for burn block to be processed");
+
+    info!("------------------------- Miner 1 Extends Tenure B -------------------------");
```
Comment (hstove) [inline testnet/stacks-node/src/tests/signer/v0.rs:7772]: Can you add an assertion that miner 2 won the sortition for block C?

File: testnet/stacks-node/src/tests/signer/v0.rs [hunk 9]
```diff
@@ -7037,3 +7037,832 @@ fn block_validation_response_timeout() {
...
+        || {
+            let stacks_height = signer_test
+                .stacks_client
+                .get_peer_info()
+                .expect("Failed to get peer info")
+                .stacks_tip_height;
+            Ok(stacks_height > stacks_height_before)
+        },
+    )
+    .expect("Timed out waiting for final block to be mined and processed");
```
Comment (hstove) [inline testnet/stacks-node/src/tests/signer/v0.rs:7858]: Same here for an assertion that miner 2 won this sortition, and maybe also a verification of a TenureChange

**Review by hstove — CHANGES_REQUESTED** [review]
Overall LGTM! Mainly requesting a change in the SignerDB migration and then a few extra assertions in your test.

File: stacks-signer/src/signerdb.rs (hunk 3 above)
Comment (obycode) [reply stacks-signer/src/signerdb.rs:418]: Yes, I thought that was overkill, but I was just following the previous code. I'll update to your method.

Comment (obycode) [reply stacks-signer/src/signerdb.rs:418]: Updated in a9acfa048.

File: testnet/stacks-node/src/nakamoto_node/relayer.rs (hunk 5 above)
Comment (obycode) [reply testnet/stacks-node/src/nakamoto_node/relayer.rs:964]: Right, we're trying to find the last good sortition, given that the current one is already bad since we arrived in `continue_tenure`. We get the last snapshot with a sortition, but if we determine that is bad, then we go back to the sortition of the tip.

File: testnet/stacks-node/src/nakamoto_node/relayer.rs (hunk 6 above)
Comment (obycode) [reply testnet/stacks-node/src/nakamoto_node/relayer.rs:944]: I do think it's still necessary. We can definitely reuse it here and avoid looking it up again.

Comment (obycode) [reply testnet/stacks-node/src/nakamoto_node/relayer.rs:944]: Hmm... maybe it's not. I'll look at this some more.

Comment (hstove) [reply testnet/stacks-node/src/nakamoto_node/relayer.rs:944]: IMO I would appreciate some ELI5(ish) comments on some of the sortdb lookups to explain why/what they're doing. There's already great comments in here, I think it's just tricky to follow when I have to constantly look up with a specific function does

Comment (obycode) [reply testnet/stacks-node/src/nakamoto_node/relayer.rs]: Ok, I tried to clean this function up a bit. What do you think about this refactoring? 21788465f

File: testnet/stacks-node/src/tests/signer/v0.rs (hunk 7 above)
Comment (obycode) [reply testnet/stacks-node/src/tests/signer/v0.rs]: d8140e063

File: testnet/stacks-node/src/tests/signer/v0.rs (hunk 8 above)
Comment (obycode) [reply testnet/stacks-node/src/tests/signer/v0.rs:7764]: 54c88c6a2

File: testnet/stacks-node/src/tests/signer/v0.rs (hunk 9 above)
Comment (obycode) [reply testnet/stacks-node/src/tests/signer/v0.rs:7850]: 54c88c6a2

File: testnet/stacks-node/src/nakamoto_node/relayer.rs (hunk 6 above)
Comment (obycode) [reply testnet/stacks-node/src/nakamoto_node/relayer.rs]: I also improved the names and added more comments in another commit just now.

**Comment by obycode** [comment]
Next step will be to remove the tenure extend disable flag and instead update those forking tests to expect this new behavior. This will be lower priority though as I pivot to the design for #5434 before coming back to this.

File: stacks-signer/src/signerdb.rs [hunk 10]
```diff
@@ -308,6 +308,10 @@ static CREATE_INDEXES_3: &str = r#"
 CREATE INDEX IF NOT EXISTS block_rejection_signer_addrs_on_block_signature_hash ON block_rejection_signer_addrs(signer_signature_hash);
 "#;
 
+static CREATE_INDEXES_4: &str = r#"
+CREATE INDEX IF NOT EXISTS blocks_state ON blocks ((json_extract(block_info, '$.state')));
```
Comment (jferrant) [inline stacks-signer/src/signerdb.rs:312]: Bit of a side note, but I think it might be best to make this query on both the state and the consensus hash since we never really ever query blocks without both.

Comment (jferrant) [reply stacks-signer/src/signerdb.rs:312]: Should we do (maybe this is overindexed...): 
```
"static CREATE_INDEXES_4: &str = r#"
CREATE INDEX IF NOT EXISTS blocks_state ON blocks (consensus_hash, json_extract(block_info, '$.state'),  stacks_height DESC);
"#;
```

**Comment by hstove** [comment]
@obycode this was the PR where it was discussed that I could help out on the integration tests, right?

**Comment by obycode** [comment]
> @obycode this was the PR where it was discussed that I could help out on the integration tests, right?

Yes, see my last comment. I had previously adjusted some tests to disable this feature, but then @kantai suggested (and I agree) that it would be better to instead adjust those forking tests to expect this behavior.

You can probably revert ba2faf7035b9e3c472d21685121c75f139f02dd2 and 965f58ba294f965741079c83d69c974310df2edd.

**Comment by hstove** [comment]
FYI I've stripped all of the shadow blocks stuff, which was merged into this PR. The old branch is now in https://github.com/stacks-network/stacks-core/tree/feat/tenure-extend-no-blocks-with-shadow

**Comment by obycode** [comment]
My last commit adds a new test, `tenure_extend_after_2_bad_commits` which will currently fail due to the signerdb issue Hank identified. That can be enabled once that problem is resolved (or determined that we shouldn't fix it).

File: testnet/stacks-node/src/tests/signer/v0.rs [hunk 11]
```diff
@@ -4334,11 +4390,36 @@ fn partial_tenure_fork() {
...
+        let tip_sn = SortitionDB::get_canonical_burn_chain_tip(sortdb.conn()).unwrap();
+        let sortition_winner = match tip_sn.miner_pk_hash {
+            Some(pk_hash) => {
+                if pk_hash == mining_pkh_1 {
+                    1
+                } else {
+                    2
+                }
+            }
+            // No sortition means miner 2 won (?)
```
Comment (obycode) [inline testnet/stacks-node/src/tests/signer/v0.rs:4402]: I think no sortition would just mean there was a problem, not that miner 2 won. There should always be a sortition.

File: testnet/stacks-node/src/tests/signer/v0.rs [hunk 12]
```diff
@@ -4435,40 +4523,48 @@ fn partial_tenure_fork() {
...
             miner_2_tenures += 1;
-            miner_2_blocks += blocks;
+        }
+
+        if miner == 1 {
+            miner_1_blocks += blocks;
+        } else {
+            // miner_2_blocks += blocks;
+            // Miner 2 should never mine any blocks
+            miner_2_blocks += 0;
```
Comment (obycode) [inline testnet/stacks-node/src/tests/signer/v0.rs:4542]: Miner 2 can still mine blocks before miner 1's first tenure, right?

File: testnet/stacks-node/src/tests/signer/v0.rs [hunk 13]
```diff
@@ -4435,40 +4523,48 @@ fn partial_tenure_fork() {
...
+        if miner == 1 && !miner_2_won_2_in_a_row {
             assert_eq!(mined_1, mined_before_1 + blocks + 1);
-        } else if miner_2_tenures < min_miner_2_tenures {
-            assert_eq!(mined_2, mined_before_2 + blocks + 1);
-        } else {
-            // Miner 2 should have mined 0 blocks after the fork
-            assert_eq!(mined_2, mined_before_2);
         }
+        // Miner 2 should have mined 0 blocks
+        assert_eq!(mined_2, mined_before_2);
```
Comment (obycode) [inline testnet/stacks-node/src/tests/signer/v0.rs:4556]: I don't follow this change. Can you explain?

File: testnet/stacks-node/src/tests/signer/v0.rs (hunk 11 above)
Comment (hstove) [reply testnet/stacks-node/src/tests/signer/v0.rs:4402]: Ah yeah pretty much all of your comments here can be addressed with "yeah, this is me trying to understand stuff and not getting it to work" lol

File: testnet/stacks-node/src/tests/signer/v0.rs [hunk 14]
```diff
@@ -4311,13 +4319,51 @@ fn partial_tenure_fork() {
...
+    // Pause block commits
+    rl1_skip_commit_op.set(true);
+    rl2_skip_commit_op.set(true);
+
+    let mined_before_1 = blocks_mined1.load(Ordering::SeqCst);
+    let mined_before_2 = blocks_mined2.load(Ordering::SeqCst);
+    let commits_before_1 = commits_1.load(Ordering::SeqCst);
+    let commits_before_2 = commits_2.load(Ordering::SeqCst);
+
+    // Ensure that both block commits have been sent before continuing
```
Comment (hstove) [inline testnet/stacks-node/src/tests/signer/v0.rs:4344]:
```suggestion
    // Mine one block before continuing
```

File: testnet/stacks-node/src/tests/signer/v0.rs [hunk 15]
```diff
@@ -4262,6 +4316,53 @@ fn partial_tenure_fork() {
...
+        &mut signer_test.running_nodes.btc_regtest_controller,
+        60,
+        || {
+            let mined_1 = blocks_mined1.load(Ordering::SeqCst);
+            let mined_2 = blocks_mined2.load(Ordering::SeqCst);
+
+            Ok(mined_1 > mined_before_1 || mined_2 > mined_before_2)
+        },
+    )
+    .expect("Timed out waiting for block commit after new Stacks block");
```
Comment (jferrant) [inline testnet/stacks-node/src/tests/signer/v0.rs:4355]: I think this expect is wrong. Should be "Timed out waiting for new Stacks block to be mined" or something similar.

File: testnet/stacks-node/src/tests/signer/v0.rs (hunk 14 above)
Comment (obycode) [reply testnet/stacks-node/src/tests/signer/v0.rs]: I had just updated this to something similar before seeing your comment.

**Review by hstove — DISMISSED** [review]
LGTM! Thanks for spending that time to figure out what was going on with the tests.

File: stacks-signer/src/signerdb.rs [hunk 16]
```diff
@@ -616,6 +640,15 @@ impl SignerDb {
         try_deserialize(result)
     }
 
+    /// Return the canonical tip -- the last globally accepted block.
+    pub fn get_canonical_tip(&self) -> Result<Option<BlockInfo>, DBError> {
```
Comment (jcnelson) [inline stacks-signer/src/signerdb.rs:644]: How do you resolve ties?  Ties happen in the wild all the time.

Comment (jferrant) [reply stacks-signer/src/signerdb.rs:644]: We should never have ties in globally accepted blocks. We can have it in locally accepted blocks, but never globally.

File: testnet/stacks-node/src/nakamoto_node/relayer.rs [hunk 17]
```diff
@@ -852,104 +877,219 @@ impl RelayerThread {
         Ok(())
     }
 
-    fn continue_tenure(&mut self, new_burn_view: ConsensusHash) -> Result<(), NakamotoNodeError> {
-        if let Err(e) = self.stop_tenure() {
-            error!("Relayer: Failed to stop tenure: {e:?}");
-            return Ok(());
+    /// Get a snapshot for an existing burn chain block given its consensus hash.
```
Comment (jcnelson) [inline testnet/stacks-node/src/nakamoto_node/relayer.rs:880]: This is a dup of `SortitionDB::get_block_snapshot_consensus()`

File: testnet/stacks-node/src/nakamoto_node/relayer.rs [hunk 18]
```diff
@@ -852,104 +877,219 @@ impl RelayerThread {
...
+                NakamotoNodeError::SnapshotNotFoundForChainTip
+            })?
+            .ok_or_else(|| {
+                error!("Relayer: failed to get block snapshot for new burn view");
+                NakamotoNodeError::SnapshotNotFoundForChainTip
+            })
+    }
+
+    /// Get the Stacks block ID for the canonical tip.
+    fn get_canonical_stacks_tip(&self) -> StacksBlockId {
```
Comment (jcnelson) [inline testnet/stacks-node/src/nakamoto_node/relayer.rs:900]: Can you instead add this to `SortitionDB` as a public method?

File: testnet/stacks-node/src/nakamoto_node/relayer.rs [hunk 19]
```diff
@@ -852,104 +877,219 @@ impl RelayerThread {
...
+            );
+            NakamotoNodeError::ParentNotFound
+        })?;
+        Ok(BlockHeaderHash(
+            highest_tenure_start_block_header.index_block_hash().0,
+        ))
+    }
+
+    /// Determine the type of tenure change to issue based on whether this
+    /// miner successfully issued a tenure change in the last tenure.
```
Comment (jcnelson) [inline testnet/stacks-node/src/nakamoto_node/relayer.rs:951]: This docstring is confusing.  The docstring implies that you're looking to see if the given miner (identified by `mining_pkh`) issued a tenure change in the last tenure, but you're not looking at `last_tenure`; you're looking instead at `canonical_tenure`.  Can you explain the difference between these two `BlockSnapshot`s?

File: testnet/stacks-node/src/nakamoto_node/relayer.rs (hunk 18 above)
Comment (hstove) [reply testnet/stacks-node/src/nakamoto_node/relayer.rs:900]: Also maybe rename to `get_canonical_stacks_tip_block_id`, though I'm not really confident on that suggestion. I think my suggestion would follow other functions like `SortitionDB::get_canonical_stacks_chain_tip_hash_and_height`

File: testnet/stacks-node/src/nakamoto_node/relayer.rs [hunk 20]
```diff
@@ -852,104 +877,219 @@ impl RelayerThread {
...
+                    error!("Relayer: failed to get block snapshot for canonical tip");
                     NakamotoNodeError::SnapshotNotFoundForChainTip
-                })?;
+                })
+        } else {
+            Ok(sn)
+        }
+    }
+
+    fn get_block_snapshot(&self, ch: &ConsensusHash) -> Result<BlockSnapshot, NakamotoNodeError> {
```
Comment (jcnelson) [inline testnet/stacks-node/src/nakamoto_node/relayer.rs:1015]: It looks like you're wrapping a lot of existing methods in `SortitionDB` and `NakamotoChainState` with private functions here, where the only real differences are in that you're logging errors and converting the errors into `NakamotoNodeError` variants.  It may be more economical to instead implement `From<ChainstateError>` for `NakamotoNodeError`, and add a text field to `NakamotoNodeError` to include the `error!()` text.

File: testnet/stacks-node/src/nakamoto_node/relayer.rs (hunk 19 above)
Comment (obycode) [reply testnet/stacks-node/src/nakamoto_node/relayer.rs:951]: Yeah, I agree this comment, and this function are a bit confusing. I don't think there was anything new here, just part of some refactoring I did to try to make things less confusing. I'll try to come up with a better comment here.

File: stacks-signer/src/signerdb.rs (hunk 16 above)
Comment (obycode) [reply stacks-signer/src/signerdb.rs:644]: Sorry, your comment is unclear @jferrant. Are you saying that we should handle a tie or that we don't need to?

I think we may have the one case where there can be two globally accepted blocks at the same height -- at the tenure boundary if miner A is slow to mine the first block and the next bitcoin block arrives.

File: testnet/stacks-node/src/nakamoto_node/relayer.rs (hunk 18 above)
Comment (obycode) [reply testnet/stacks-node/src/nakamoto_node/relayer.rs]: Added in c2131150e636190c756e6e695d44a544dd419daa

File: testnet/stacks-node/src/nakamoto_node/relayer.rs (hunk 19 above)
Comment (obycode) [reply testnet/stacks-node/src/nakamoto_node/relayer.rs]: How's this? 2e963e7c80b034207b9cfff7d4aa20cfb30faaf8

File: testnet/stacks-node/src/nakamoto_node/relayer.rs (hunk 20 above)
Comment (obycode) [reply testnet/stacks-node/src/nakamoto_node/relayer.rs:1008]: Yeah, the wrapper functions were part of my effort to try to clean up `continue_tenure` because it was getting really big and confusing. I will give this a try to see if it helps.

File: stacks-signer/src/signerdb.rs (hunk 16 above)
Comment (jferrant) [reply stacks-signer/src/signerdb.rs:644]: I am agreeing with Jude. For the reason you mention, at the boundary, we could technically reorg the last block so we cannot have a canonical tip known without the context of a consensus hash.  

My comment is asking if we HAVE to know the canonical tip or if knowing just the last globally accepted block for a given consensus hash is enough (avoid the need for a tie breaker). If not, we will have to have a tie breaker as Jude mentions to accommodate the potential reorg at the boundary that you mention.

File: testnet/stacks-node/src/nakamoto_node/relayer.rs (hunk 19 above)
Comment (jcnelson) [reply testnet/stacks-node/src/nakamoto_node/relayer.rs]: Better. Thanks!

File: testnet/stacks-node/src/nakamoto_node/relayer.rs (hunk 18 above)
Comment (jcnelson) [reply testnet/stacks-node/src/nakamoto_node/relayer.rs]: Thanks!

File: stacks-signer/src/signerdb.rs (hunk 16 above)
Comment (hstove) [reply stacks-signer/src/signerdb.rs:644]: Yeah I follow the issue.

I don't think we can take a consensus hash for this function, because the point of this function is to determine whether a block proposal's consensus hash is valid.

I do think using the last globally accepted block (by using `signed_group` as a tie breaker) works here, but I'm open to feedback. I'll add the `signed_group` tiebreaker for now, but happy to take a different approach.

Comment (hstove) [reply stacks-signer/src/signerdb.rs:668]: Addressed in [c8f1d99](https://github.com/stacks-network/stacks-core/pull/5452/commits/c8f1d999d6f6520b5a1a68ec237f215529351e48)

File: testnet/stacks-node/src/nakamoto_node/relayer.rs [hunk 21]
```diff
@@ -1032,7 +995,12 @@ impl RelayerThread {
         debug!("Relayer: successfully stopped tenure.");
 
         // Get the necessary snapshots and state
-        let burn_tip = self.get_block_snapshot_consensus(&new_burn_view)?;
+        let burn_tip =
+            SortitionDB::get_block_snapshot_consensus(self.sortdb.conn(), &new_burn_view)?
+                .ok_or_else(|| {
+                    error!("Relayer: failed to get block snapshot for new burn view");
+                    NakamotoNodeError::SnapshotNotFoundForChainTip
+                })?;
```
Comment (hstove) [inline testnet/stacks-node/src/nakamoto_node/relayer.rs:1003]: In this case, I removed the wrapper function, and I also removed the `map_err`. The `map_err` is only for `DBError`, which IMO deserves a different `NakamotoNodeError` (which I've added) and (again IMO) doesn't need a custom error log.

Comment (hstove) [reply testnet/stacks-node/src/nakamoto_node/relayer.rs:1003]: (Same exact story with a few lines below)

File: testnet/stacks-node/src/nakamoto_node/miner.rs [hunk 22]
```diff
@@ -1082,13 +1077,11 @@ impl BlockMinerThread {
             ) {
                 error!("Relayer: Failure mining anchored block: {e}");
             }
-            NakamotoNodeError::MiningFailure(e)
+            e
```
Comment (hstove) [inline testnet/stacks-node/src/nakamoto_node/miner.rs:1080]: Unrelated to this PR, but the 3 lines above - why not just always have that `error!` log?

File: testnet/stacks-node/src/nakamoto_node.rs [hunk 23]
```diff
@@ -71,46 +73,73 @@ pub struct StacksNode {
...
+    #[error("A new parent has been discovered since mining started")]
     NewParentDiscovered,
     /// A failure occurred while constructing a VRF Proof
+    #[error("A failure occurred while constructing a VRF Proof")]
     BadVrfConstruction,
+    #[error("The miner didn't accept their own block")]
     CannotSelfSign,
-    MiningFailure(ChainstateError),
+    #[error("A failure occurred while mining: {0}")]
+    MiningFailure(#[from] ChainstateError),
```
Comment (hstove) [inline testnet/stacks-node/src/nakamoto_node.rs:120]: Note that we have two error variants that wrap `ChainstateError` (this and `AcceptFailure`). I've added the `#[from]` to this one because we almost always use this variant - we only use `AcceptFailure` in one case.

File: testnet/stacks-node/src/nakamoto_node/miner.rs (hunk 22 above)
Comment (obycode) [reply testnet/stacks-node/src/nakamoto_node/miner.rs:1080]: I think that's because those two types of errors are normal and happen often, so we don't want to report an error in those cases.

**Review by jcnelson — APPROVED** [review]

File: stacks-signer/src/signerdb.rs [hunk 24]
```diff
@@ -421,9 +429,19 @@ static SCHEMA_3: &[&str] = &[
     "INSERT INTO db_config (version) VALUES (3);",
 ];
 
+static SCHEMA_4: &[&str] = &[
```
Comment (jferrant) [inline stacks-signer/src/signerdb.rs:432]: I think you can make this a single schema now! I know it was merged in from another PR but that PR has yet to hit develop so why don't we combine these?

**Review by jferrant — CHANGES_REQUESTED** [review]
Really just the one nit about combining the schemas to reduce the number of unnecessary versions. :)

File: stacks-signer/src/signerdb.rs (hunk 24 above)
Comment (obycode) [reply stacks-signer/src/signerdb.rs:432]: Yeah, good point. I'll make that change.

Comment (obycode) [reply stacks-signer/src/signerdb.rs:429]: Fixed in ecae254ef.

**Review by hstove — APPROVED** [review]

**Review by jcnelson — APPROVED** [review]

**Review by jferrant — APPROVED** [review]
